File: src/features/users/users.service.ts
import { BadRequestException, ConflictException, ForbiddenException, Injectable, NotFoundException } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { ClientSession, Model } from 'mongoose';
import * as argon2 from 'argon2';
import { AuditLogService } from '../../common/services/audit-log.service';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { Role, canAssignRoles, mergeRoles, resolvePrimaryRole } from '../../common/roles';
import { User } from './schemas/user.schema';
import { STRONG_PASSWORD_REGEX } from '../auth/dto/password-rules';
import { SeatsService } from '../seats/seats.service';
import { seatConfig } from '../../config/app.config';

export type ActorContext = { role?: Role; roles?: Role[]; orgId?: string; id?: string };

@Injectable()
export class UsersService {
  constructor(
    @InjectModel('User') private readonly userModel: Model<User>,
    private readonly audit: AuditLogService,
    private readonly seats: SeatsService
  ) {}

  private sanitize(user: any) {
    const {
      passwordHash,
      verificationTokenHash,
      verificationTokenExpires,
      resetTokenHash,
      resetTokenExpires,
      ...rest
    } = user.toObject ? user.toObject() : user;
    return rest;
  }

  private summarizeDiff(before: any, after: any, fields: string[]) {
    const changes: Record<string, { from: any; to: any }> = {};
    fields.forEach((field) => {
      if (before[field] === after[field]) return;
      changes[field] = { from: before[field], to: after[field] };
    });
    return changes;
  }

  private normalizeRoles(dto: Pick<CreateUserDto, 'role' | 'roles'>) {
    const roles = mergeRoles(dto.role as Role, dto.roles as Role[]);
    const primary = resolvePrimaryRole(roles);
    return { roles, primary };
  }

  private getActorRoles(actor: ActorContext) {
    return mergeRoles(actor.role as Role, actor.roles as Role[]);
  }

  private assertOrgScope(userOrgId: string | undefined | null, actor: ActorContext) {
    const actorRoles = this.getActorRoles(actor);
    const canBypass = actorRoles.includes(Role.SuperAdmin) || actorRoles.includes(Role.PlatformAdmin);
    if (canBypass) return;
    if (!actor.orgId) {
      throw new ForbiddenException('Missing organization context');
    }
    if (userOrgId && actor.orgId !== userOrgId) {
      throw new ForbiddenException('Cannot modify user outside your organization');
    }
  }

  private validateRoleChange(actor: ActorContext, roles: Role[]) {
    const normalizedRoles = mergeRoles(undefined, roles);
    if (!normalizedRoles.length) throw new BadRequestException('At least one role is required');
    const actorRoles = this.getActorRoles(actor);
    if (!actorRoles.length) {
      throw new ForbiddenException('Actor has no roles assigned');
    }
    if (!canAssignRoles(actorRoles, normalizedRoles)) {
      throw new ForbiddenException('Insufficient role to assign requested roles');
    }
    return normalizedRoles;
  }

  private resolveOrg(actor: ActorContext, requestedOrgId?: string) {
    const actorRoles = this.getActorRoles(actor);
    const privileged = actorRoles.includes(Role.SuperAdmin) || actorRoles.includes(Role.PlatformAdmin);
    const orgId = privileged && requestedOrgId ? requestedOrgId : actor.orgId;
    if (!orgId) throw new ForbiddenException('Missing organization context');
    return { orgId, actorRoles, privileged };
  }

  private canManageCompliance(actorRoles: Role[]) {
    return actorRoles.some((role) =>
      [
        Role.SuperAdmin,
        Role.PlatformAdmin,
        Role.Admin,
        Role.OrgOwner,
        Role.OrgAdmin,
        Role.Compliance,
        Role.ComplianceOfficer,
      ].includes(role)
    );
  }

  private assertNotLegalHold(user: any, action: string) {
    if (user.legalHold) {
      throw new ForbiddenException(`Cannot ${action} user on legal hold`);
    }
  }

  async create(dto: CreateUserDto, actor?: ActorContext, options?: { session?: ClientSession; enforceSeat?: boolean }) {
    const email = (dto.email || '').trim().toLowerCase();
    if (!email) throw new BadRequestException('Email is required');
    const orgId = dto.orgId;
    if (!orgId) throw new ConflictException('orgId is required');

    const enforceSeat = options?.enforceSeat ?? false;
    if (enforceSeat && !options?.session) {
      await this.seats.ensureOrgSeats(orgId, seatConfig.defaultSeatsPerOrg);
    }

    if (enforceSeat && !options?.session) {
      const session = await this.userModel.db.startSession();
      try {
        session.startTransaction();
        const result = await this.create(dto, actor, { ...options, session });
        await session.commitTransaction();
        return result;
      } catch (err) {
        await session.abortTransaction();
        throw err;
      } finally {
        session.endSession();
      }
    }

    const session = options?.session;
    const existingQuery = this.userModel.findOne({ email });
    if (session) existingQuery.session(session);
    const existing = await existingQuery;
    if (existing) throw new ConflictException('Email already registered');
    if (!STRONG_PASSWORD_REGEX.test(dto.password)) {
      throw new BadRequestException('Password does not meet strength requirements');
    }
    const { roles, primary } = this.normalizeRoles(dto);
    if (actor) {
      this.validateRoleChange(actor, roles);
    }
    const passwordHash = await argon2.hash(dto.password, { type: argon2.argon2id });
    const payload = {
      username: dto.username,
      email,
      passwordHash,
      role: primary,
      roles,
      orgId,
      isEmailVerified: dto.isEmailVerified ?? false,
      verificationTokenHash: dto.verificationTokenHash ?? null,
      verificationTokenExpires: dto.verificationTokenExpires ?? null,
      resetTokenHash: dto.resetTokenHash ?? null,
      resetTokenExpires: dto.resetTokenExpires ?? null,
      isOrgOwner: dto.isOrgOwner ?? roles.includes(Role.OrgOwner),
      archivedAt: null,
      lastLogin: null,
      piiStripped: false,
      legalHold: false,
    };

    let user: any;
    if (session) {
      const created = await (this.userModel as any).create([payload], { session });
      user = Array.isArray(created) ? created[0] : created;
    } else {
      user = await (this.userModel as any).create(payload);
    }

    if (enforceSeat) {
      await this.seats.allocateSeat(orgId, user.id, { role: primary, session });
    }

    await this.audit.log({
      eventType: 'user.created',
      userId: actor?.id || user.id,
      orgId: user.orgId,
      entity: 'User',
      entityId: user.id,
    }, session ? { session } : undefined);
    return this.sanitize(user);
  }

  async findByEmail(email: string) {
    const normalized = (email || '').trim().toLowerCase();
    if (!normalized) return null;
    return this.userModel.findOne({ email: normalized, archivedAt: null });
  }

  async findAnyByEmail(email: string) {
    const normalized = (email || '').trim().toLowerCase();
    if (!normalized) return null;
    return this.userModel.findOne({ email: normalized }).lean();
  }

  async scopedFindAll(orgId: string) {
    const users = await this.userModel.find({ orgId, archivedAt: null });
    return users.map((u) => this.sanitize(u));
  }

  async list(
    actor: ActorContext,
    options: {
      orgId?: string;
      includeArchived?: boolean;
    } = {}
  ) {
    const { orgId } = this.resolveOrg(actor, options.orgId);
    const filter: any = { orgId };
    if (!options.includeArchived) {
      filter.archivedAt = null;
    }
    const users = await this.userModel.find(filter);
    return users.map((u) => this.sanitize(u));
  }

  async getById(id: string, actor: ActorContext, includeArchived = false) {
    const user = await this.userModel.findById(id);
    if (!user) throw new NotFoundException('User not found');
    this.assertOrgScope(user.orgId, actor);
    if (user.archivedAt && !includeArchived) {
      throw new NotFoundException('User not found');
    }
    return this.sanitize(user);
  }

  async getByIdForSession(id: string) {
    const user = await this.userModel.findById(id);
    if (!user) throw new NotFoundException('User not found');
    return this.sanitize(user);
  }

  async findByVerificationToken(hash: string) {
    return this.userModel.findOne({
      verificationTokenHash: hash,
      verificationTokenExpires: { $gt: new Date() },
      archivedAt: null,
      legalHold: { $ne: true },
    });
  }

  async setVerificationToken(userId: string, hash: string, expires: Date) {
    return this.userModel.findByIdAndUpdate(
      userId,
      { verificationTokenHash: hash, verificationTokenExpires: expires, isEmailVerified: false },
      { new: true }
    );
  }

  async clearVerificationToken(userId: string) {
    return this.userModel.findByIdAndUpdate(
      userId,
      { verificationTokenHash: null, verificationTokenExpires: null, isEmailVerified: true },
      { new: true }
    );
  }

  async findByResetToken(hash: string) {
    return this.userModel.findOne({
      resetTokenHash: hash,
      resetTokenExpires: { $gt: new Date() },
      archivedAt: null,
      legalHold: { $ne: true },
    });
  }

  async setResetToken(userId: string, hash: string, expires: Date) {
    return this.userModel.findByIdAndUpdate(
      userId,
      { resetTokenHash: hash, resetTokenExpires: expires },
      { new: true }
    );
  }

  async clearResetTokenAndSetPassword(userId: string, newPassword: string) {
    const passwordHash = await argon2.hash(newPassword, { type: argon2.argon2id });
    return this.userModel.findByIdAndUpdate(
      userId,
      {
        passwordHash,
        resetTokenHash: null,
        resetTokenExpires: null,
      },
      { new: true }
    );
  }

  async archiveUser(targetUserId: string, actor: ActorContext) {
    const user = await this.userModel.findById(targetUserId);
    if (!user) throw new NotFoundException('User not found');
    this.assertOrgScope(user.orgId, actor);
    this.assertNotLegalHold(user, 'archive');
    if (user.archivedAt) return this.sanitize(user);
    user.archivedAt = new Date();
    await user.save();
    await this.seats.releaseSeatForUser(user.orgId as string, user.id);
    await this.audit.log({
      eventType: 'user.archived',
      userId: actor.id || user.id,
      orgId: user.orgId,
      entity: 'User',
      entityId: user.id,
      metadata: { archivedAt: user.archivedAt },
    });
    return this.sanitize(user);
  }

  async unarchiveUser(targetUserId: string, actor: ActorContext) {
    const user = await this.userModel.findById(targetUserId);
    if (!user) throw new NotFoundException('User not found');
    this.assertOrgScope(user.orgId, actor);
    this.assertNotLegalHold(user, 'unarchive');
    if (!user.archivedAt) return this.sanitize(user);

    const orgId = user.orgId as string;
    await this.seats.ensureOrgSeats(orgId, seatConfig.defaultSeatsPerOrg);
    await this.seats.allocateSeat(orgId, user.id, { role: user.role });

    try {
      user.archivedAt = null;
      await user.save();
    } catch (err) {
      await this.seats.releaseSeatForUser(orgId, user.id);
      throw err;
    }
    await this.audit.log({
      eventType: 'user.unarchived',
      userId: actor.id || user.id,
      orgId: user.orgId,
      entity: 'User',
      entityId: user.id,
      metadata: { archivedAt: user.archivedAt },
    });
    return this.sanitize(user);
  }

  async getUserRoles(targetUserId: string, actor: ActorContext) {
    const user = await this.userModel.findById(targetUserId);
    if (!user) throw new NotFoundException('User not found');
    this.assertOrgScope(user.orgId, actor);
    const sanitized = this.sanitize(user) as any;
    const roles = mergeRoles(sanitized.role as Role, sanitized.roles as Role[]);
    return { ...sanitized, roles, role: sanitized.role || resolvePrimaryRole(roles) };
  }

  async listOrgRoles(orgId: string | undefined, actor: ActorContext) {
    const actorRoles = this.getActorRoles(actor);
    const canBypass = actorRoles.includes(Role.SuperAdmin) || actorRoles.includes(Role.PlatformAdmin);
    if (!canBypass && orgId && actor.orgId && orgId !== actor.orgId) {
      throw new ForbiddenException('Cannot access users outside your organization');
    }
    const resolvedOrg = canBypass && orgId ? orgId : actor.orgId;
    if (!resolvedOrg) throw new ForbiddenException('Missing organization context');
    const users = await this.userModel.find({ orgId: resolvedOrg, archivedAt: null });
    return users.map((u) => {
      const sanitized = this.sanitize(u) as any;
      const roles = mergeRoles(sanitized.role as Role, sanitized.roles as Role[]);
      return { ...sanitized, roles, role: sanitized.role || resolvePrimaryRole(roles) };
    });
  }

  async updateRoles(targetUserId: string, roles: Role[], actor: ActorContext) {
    const user = await this.userModel.findById(targetUserId);
    if (!user) throw new NotFoundException('User not found');
    this.assertOrgScope(user.orgId, actor);
    const normalizedRoles = this.validateRoleChange(actor, roles);
    if (user.archivedAt) {
      throw new ForbiddenException('Cannot change roles for archived users');
    }
    if (user.legalHold) {
      throw new ForbiddenException('Cannot change roles while user is on legal hold');
    }
    user.roles = normalizedRoles;
    user.role = resolvePrimaryRole(normalizedRoles);
    user.isOrgOwner = normalizedRoles.includes(Role.OrgOwner);
    await user.save();
    await this.audit.log({
      eventType: 'user.roles_updated',
      userId: actor.id || user.id,
      orgId: user.orgId,
      entity: 'User',
      entityId: user.id,
      metadata: { roles: normalizedRoles, actorRoles: this.getActorRoles(actor) },
    });
    return this.sanitize(user);
  }

  async update(id: string, dto: UpdateUserDto, actor: ActorContext) {
    const user = await this.userModel.findById(id);
    if (!user) throw new NotFoundException('User not found');
    this.assertOrgScope(user.orgId, actor);
    if (user.archivedAt) {
      throw new ForbiddenException('Cannot update archived users');
    }
    const actorRoles = this.getActorRoles(actor);
    const canManageCompliance = this.canManageCompliance(actorRoles);
    if (user.legalHold && !canManageCompliance) {
      throw new ForbiddenException('User on legal hold');
    }
    const before = this.sanitize(user);

    if (dto.email && dto.email !== user.email) {
      const normalizedEmail = dto.email.toLowerCase();
      const duplicate = await this.userModel.findOne({ email: normalizedEmail });
      if (duplicate && duplicate.id !== id) {
        throw new ConflictException('Email already registered');
      }
      user.email = normalizedEmail;
    }
    if (dto.username !== undefined) user.username = dto.username;
    if (dto.isEmailVerified !== undefined) {
      if (!canManageCompliance) {
        throw new ForbiddenException('Insufficient role to change verification state');
      }
      user.isEmailVerified = dto.isEmailVerified;
    }
    if (dto.piiStripped !== undefined) {
      if (!canManageCompliance) throw new ForbiddenException('Insufficient role to change compliance fields');
      user.piiStripped = dto.piiStripped;
    }
    if (dto.legalHold !== undefined) {
      if (!canManageCompliance) throw new ForbiddenException('Insufficient role to change compliance fields');
      user.legalHold = dto.legalHold;
    }

    await user.save();
    const after = this.sanitize(user);
    await this.audit.log({
      eventType: 'user.updated',
      userId: actor.id || user.id,
      orgId: user.orgId,
      entity: 'User',
      entityId: user.id,
      metadata: { changes: this.summarizeDiff(before, after, ['username', 'email', 'isEmailVerified', 'piiStripped', 'legalHold']) },
    });
    return after;
  }

  async markLastLogin(userId: string) {
    return this.userModel.findByIdAndUpdate(
      userId,
      { lastLogin: new Date() },
      { new: true, lean: true }
    );
  }
}

File: spring-boot-api/src/main/java/com/mytegroup/api/service/users/UsersService.java
package com.mytegroup.api.service.users;

import com.mytegroup.api.common.enums.Role;
import com.mytegroup.api.common.util.PasswordValidator;
import com.mytegroup.api.entity.core.Organization;
import com.mytegroup.api.entity.core.User;
import com.mytegroup.api.exception.BadRequestException;
import com.mytegroup.api.exception.ConflictException;
import com.mytegroup.api.exception.ForbiddenException;
import com.mytegroup.api.exception.ResourceNotFoundException;
import com.mytegroup.api.repository.core.OrganizationRepository;
import com.mytegroup.api.repository.core.UserRepository;
import com.mytegroup.api.service.common.AuditLogService;
import com.mytegroup.api.service.common.RoleExpansionHelper;
import com.mytegroup.api.service.common.ServiceAuthorizationHelper;
import com.mytegroup.api.service.seats.SeatsService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * Service for user management.
 * Handles CRUD operations, role management, email verification, and password reset.
 * Mirrors NestJS users.service.ts implementation.
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class UsersService {
    
    private final UserRepository userRepository;
    private final OrganizationRepository organizationRepository;
    private final PasswordEncoder passwordEncoder;
    private final AuditLogService auditLogService;
    private final ServiceAuthorizationHelper authHelper;
    private final SeatsService seatsService;
    
    @Value("${app.seats.default-per-org:5}")
    private int defaultSeatsPerOrg;
    
    /**
     * Creates a new user with optional seat enforcement.
     */
    @Transactional
    public User create(User user, boolean enforceSeat) {
        String email = normalizeEmail(user.getEmail());
        if (email == null || email.isEmpty()) {
            throw new BadRequestException("Email is required");
        }
        
        // Check for email collision
        if (userRepository.existsByEmail(email)) {
            throw new ConflictException("Email already registered");
        }
        
        // Validate organization exists
        if (user.getOrganization() == null || user.getOrganization().getId() == null) {
            throw new BadRequestException("Organization is required");
        }
        
        Long orgId = user.getOrganization().getId();
        
        // Ensure seats exist if enforcing
        if (enforceSeat) {
            seatsService.ensureOrgSeats(orgId.toString(), defaultSeatsPerOrg);
        }
        
        // Validate password strength if provided
        if (user.getPasswordHash() != null && !user.getPasswordHash().isEmpty()) {
            if (!PasswordValidator.isStrong(user.getPasswordHash())) {
                throw new BadRequestException(PasswordValidator.STRONG_PASSWORD_MESSAGE);
            }
            user.setPasswordHash(passwordEncoder.encode(user.getPasswordHash()));
        }
        
        // Normalize email
        user.setEmail(email);
        
        // Normalize and merge roles
        List<Role> roles = normalizeRoles(user.getRole(), user.getRoles());
        Role primaryRole = resolvePrimaryRole(roles);
        
        // Role validation will be handled when sessions are implemented
        user.setRole(primaryRole);
        user.setRoles(roles);
        
        // Set defaults
        if (user.getIsEmailVerified() == null) {
            user.setIsEmailVerified(false);
        }
        if (user.getIsOrgOwner() == null) {
            user.setIsOrgOwner(roles.contains(Role.ORG_OWNER));
        }
        if (user.getPiiStripped() == null) {
            user.setPiiStripped(false);
        }
        if (user.getLegalHold() == null) {
            user.setLegalHold(false);
        }
        
        User savedUser = userRepository.save(user);
        
        // Allocate seat if enforcing
        if (enforceSeat) {
            seatsService.allocateSeat(orgId.toString(), savedUser.getId(), primaryRole.getValue(), null);
        }
        
        auditLogService.log(
            "user.created",
            orgId.toString(),
            null, // userId will be set when sessions are implemented
            "User",
            savedUser.getId().toString(),
            null
        );
        
        return savedUser;
    }
    
    /**
     * Creates a new user (no seat enforcement).
     */
    @Transactional
    public User create(User user) {
        return create(user, false);
    }
    
    /**
     * Finds a user by email (active only).
     */
    @Transactional(readOnly = true)
    public User findByEmail(String email) {
        String normalized = normalizeEmail(email);
        if (normalized == null || normalized.isEmpty()) {
            return null;
        }
        return userRepository.findByEmail(normalized)
            .filter(user -> user.getArchivedAt() == null)
            .orElse(null);
    }
    
    /**
     * Finds a user by email (including archived).
     */
    @Transactional(readOnly = true)
    public User findAnyByEmail(String email) {
        String normalized = normalizeEmail(email);
        if (normalized == null || normalized.isEmpty()) {
            return null;
        }
        return userRepository.findByEmail(normalized).orElse(null);
    }
    
    /**
     * Lists users for an organization.
     */
    @Transactional(readOnly = true)
    public List<User> list(String orgId, boolean includeArchived) {
        if (orgId == null) {
            throw new BadRequestException("orgId is required");
        }
        Long orgIdLong = Long.parseLong(orgId);
        
        if (includeArchived) {
            return userRepository.findByOrgId(orgIdLong);
        } else {
            return userRepository.findByOrgIdAndArchivedAtIsNull(orgIdLong, 
                org.springframework.data.domain.PageRequest.of(0, Integer.MAX_VALUE)).getContent();
        }
    }
    
    /**
     * Gets a user by ID.
     */
    @Transactional(readOnly = true)
    public User getById(Long id, boolean includeArchived) {
        User user = userRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        
        if (!includeArchived && user.getArchivedAt() != null) {
            throw new ResourceNotFoundException("User not found");
        }
        
        return user;
    }
    
    /**
     * Gets a user by ID (for session, no auth check).
     */
    @Transactional(readOnly = true)
    public User getByIdForSession(Long id) {
        return userRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));
    }
    
    /**
     * Finds user by verification token hash.
     * @throws BadRequestException if token is invalid or expired
     */
    @Transactional(readOnly = true)
    public User findByVerificationToken(String hash) {
        return userRepository.findByVerificationTokenHash(hash, LocalDateTime.now())
            .orElseThrow(() -> new BadRequestException("Invalid or expired verification token"));
    }
    
    /**
     * Sets verification token for a user.
     */
    @Transactional
    public void setVerificationToken(Long userId, String hash, LocalDateTime expires) {
        userRepository.setVerificationToken(userId, hash, expires);
    }
    
    /**
     * Clears verification token and marks email as verified.
     */
    @Transactional
    public User clearVerificationToken(Long userId) {
        userRepository.clearVerificationToken(userId);
        return userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));
    }
    
    /**
     * Finds user by password reset token hash.
     * @throws BadRequestException if token is invalid or expired
     */
    @Transactional(readOnly = true)
    public User findByResetToken(String hash) {
        return userRepository.findByResetTokenHash(hash, LocalDateTime.now())
            .orElseThrow(() -> new BadRequestException("Invalid or expired reset token"));
    }
    
    /**
     * Sets password reset token for a user.
     */
    @Transactional
    public void setResetToken(Long userId, String hash, LocalDateTime expires) {
        userRepository.setResetToken(userId, hash, expires);
    }
    
    /**
     * Clears reset token and sets new password.
     */
    @Transactional
    public User clearResetTokenAndSetPassword(Long userId, String newPassword) {
        if (!PasswordValidator.isStrong(newPassword)) {
            throw new BadRequestException(PasswordValidator.STRONG_PASSWORD_MESSAGE);
        }
        String passwordHash = passwordEncoder.encode(newPassword);
        userRepository.clearResetTokenAndSetPassword(userId, passwordHash);
        return userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));
    }
    
    /**
     * Marks last login timestamp.
     * @throws ResourceNotFoundException if user not found
     */
    @Transactional
    public User markLastLogin(Long userId) {
        userRepository.updateLastLogin(userId, LocalDateTime.now());
        return userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));
    }
    
    /**
     * Updates a user.
     */
    @Transactional
    public User update(Long id, User userUpdates) {
        User user = userRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        
        if (user.getArchivedAt() != null) {
            throw new ForbiddenException("Cannot update archived users");
        }
        
        // Compliance checks will be handled when sessions are implemented
        boolean canManageCompliance = true; // TODO: Implement when sessions are added
        
        if (Boolean.TRUE.equals(user.getLegalHold()) && !canManageCompliance) {
            throw new ForbiddenException("User on legal hold");
        }
        
        Map<String, Object> changes = new HashMap<>();
        
        // Update email if provided
        if (userUpdates.getEmail() != null && !userUpdates.getEmail().equals(user.getEmail())) {
            String normalizedEmail = normalizeEmail(userUpdates.getEmail());
            if (userRepository.findByEmail(normalizedEmail)
                .filter(u -> !u.getId().equals(id))
                .isPresent()) {
                throw new ConflictException("Email already registered");
            }
            changes.put("email", Map.of("from", user.getEmail(), "to", normalizedEmail));
            user.setEmail(normalizedEmail);
        }
        
        // Update username if provided
        if (userUpdates.getUsername() != null) {
            changes.put("username", Map.of("from", user.getUsername(), "to", userUpdates.getUsername()));
            user.setUsername(userUpdates.getUsername());
        }
        
        // Update email verification status (requires compliance role)
        if (userUpdates.getIsEmailVerified() != null) {
            if (!canManageCompliance) {
                throw new ForbiddenException("Insufficient role to change verification state");
            }
            changes.put("isEmailVerified", Map.of("from", user.getIsEmailVerified(), "to", userUpdates.getIsEmailVerified()));
            user.setIsEmailVerified(userUpdates.getIsEmailVerified());
        }
        
        // Update compliance fields
        if (userUpdates.getPiiStripped() != null) {
            if (!canManageCompliance) {
                throw new ForbiddenException("Insufficient role to change compliance fields");
            }
            changes.put("piiStripped", Map.of("from", user.getPiiStripped(), "to", userUpdates.getPiiStripped()));
            user.setPiiStripped(userUpdates.getPiiStripped());
        }
        
        if (userUpdates.getLegalHold() != null) {
            if (!canManageCompliance) {
                throw new ForbiddenException("Insufficient role to change compliance fields");
            }
            changes.put("legalHold", Map.of("from", user.getLegalHold(), "to", userUpdates.getLegalHold()));
            user.setLegalHold(userUpdates.getLegalHold());
        }
        
        User savedUser = userRepository.save(user);
        
        Map<String, Object> metadata = new HashMap<>();
        metadata.put("changes", changes);
        
        auditLogService.log(
            "user.updated",
            savedUser.getOrganization() != null ? savedUser.getOrganization().getId().toString() : null,
            null, // userId will be set when sessions are implemented
            "User",
            savedUser.getId().toString(),
            metadata
        );
        
        return savedUser;
    }
    
    /**
     * Updates user roles.
     */
    @Transactional
    public User updateRoles(Long userId, List<Role> newRoles) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        
        if (user.getArchivedAt() != null) {
            throw new ForbiddenException("Cannot change roles for archived users");
        }
        
        if (Boolean.TRUE.equals(user.getLegalHold())) {
            throw new ForbiddenException("Cannot change roles while user is on legal hold");
        }
        
        List<Role> normalizedRoles = normalizeRoles(null, newRoles);
        // Role validation will be handled when sessions are implemented
        
        Role primaryRole = resolvePrimaryRole(normalizedRoles);
        
        user.setRoles(normalizedRoles);
        user.setRole(primaryRole);
        user.setIsOrgOwner(normalizedRoles.contains(Role.ORG_OWNER));
        
        User savedUser = userRepository.save(user);
        
        Map<String, Object> metadata = new HashMap<>();
        metadata.put("roles", normalizedRoles);
        
        auditLogService.log(
            "user.roles_updated",
            savedUser.getOrganization() != null ? savedUser.getOrganization().getId().toString() : null,
            null, // userId will be set when sessions are implemented
            "User",
            savedUser.getId().toString(),
            metadata
        );
        
        return savedUser;
    }
    
    /**
     * Archives a user.
     */
    @Transactional
    public User archive(Long id) {
        User user = userRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        
        authHelper.ensureNotOnLegalHold(user, "archive");
        
        if (user.getArchivedAt() != null) {
            return user;
        }
        
        user.setArchivedAt(LocalDateTime.now());
        User savedUser = userRepository.save(user);
        
        // Release seat
        if (user.getOrganization() != null) {
            seatsService.releaseSeatForUser(user.getOrganization().getId().toString(), savedUser.getId());
        }
        
        Map<String, Object> metadata = new HashMap<>();
        metadata.put("archivedAt", savedUser.getArchivedAt());
        
        auditLogService.log(
            "user.archived",
            savedUser.getOrganization() != null ? savedUser.getOrganization().getId().toString() : null,
            null, // userId will be set when sessions are implemented
            "User",
            savedUser.getId().toString(),
            metadata
        );
        
        return savedUser;
    }
    
    /**
     * Unarchives a user.
     */
    @Transactional
    public User unarchive(Long id) {
        User user = userRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        
        authHelper.ensureNotOnLegalHold(user, "unarchive");
        
        if (user.getArchivedAt() == null) {
            return user;
        }
        
        // Ensure seats and allocate
        if (user.getOrganization() != null) {
            String orgId = user.getOrganization().getId().toString();
            seatsService.ensureOrgSeats(orgId, defaultSeatsPerOrg);
            seatsService.allocateSeat(orgId, user.getId(), user.getRole().getValue(), null);
        }
        
        user.setArchivedAt(null);
        User savedUser = userRepository.save(user);
        
        Map<String, Object> metadata = new HashMap<>();
        metadata.put("archivedAt", savedUser.getArchivedAt());
        
        auditLogService.log(
            "user.unarchived",
            savedUser.getOrganization() != null ? savedUser.getOrganization().getId().toString() : null,
            null, // userId will be set when sessions are implemented
            "User",
            savedUser.getId().toString(),
            metadata
        );
        
        return savedUser;
    }
    
    /**
     * Gets user roles with expansion.
     */
    @Transactional(readOnly = true)
    public Map<String, Object> getUserRoles(Long userId) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        
        List<Role> roles = normalizeRoles(user.getRole(), user.getRoles());
        
        Map<String, Object> result = new HashMap<>();
        result.put("id", user.getId());
        result.put("email", user.getEmail());
        result.put("username", user.getUsername());
        result.put("roles", roles);
        result.put("role", user.getRole() != null ? user.getRole() : resolvePrimaryRole(roles));
        
        return result;
    }
    
    // ==================== Helper Methods ====================
    
    private String normalizeEmail(String email) {
        if (email == null) {
            return null;
        }
        return email.trim().toLowerCase();
    }
    
    private List<Role> normalizeRoles(Role primary, List<Role> roles) {
        List<Role> merged = new ArrayList<>();
        if (roles != null) {
            merged.addAll(roles);
        }
        if (primary != null && !merged.contains(primary)) {
            merged.add(primary);
        }
        // Remove nulls and duplicates
        merged = merged.stream()
            .distinct()
            .toList();
        
        if (merged.isEmpty()) {
            merged = new ArrayList<>(List.of(Role.USER));
        }
        return new ArrayList<>(merged);
    }
    
    private Role resolvePrimaryRole(List<Role> roles) {
        List<Role> priority = RoleExpansionHelper.getRolePriority();
        for (Role role : priority) {
            if (roles.contains(role)) {
                return role;
            }
        }
        return Role.USER;
    }
    
    // TODO: Re-implement validateRoleChange when sessions are implemented
    // private void validateRoleChange(List<Role> targetRoles) {
    //     // Role validation will be handled when sessions are implemented
    // }
    
    private boolean canManageCompliance(List<Role> actorRoles) {
        return actorRoles.stream().anyMatch(role ->
            role == Role.SUPER_ADMIN || role == Role.PLATFORM_ADMIN || role == Role.ADMIN ||
            role == Role.ORG_OWNER || role == Role.ORG_ADMIN || role == Role.COMPLIANCE ||
            role == Role.COMPLIANCE_OFFICER
        );
    }
}
